### DISCUSSION TITLE
CCL Feedback

### DISCUSSION THREAD

**Discussion by SimonMerrett** [SpenceKonde/megaTinyCore#744] [Library discussion]
So this may not be in the correct category but I'm looking at an example from the logic library, specifically [LatchNoSeq.ino](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/libraries/Logic/examples/LatchNoSeq/LatchNoSeq.ino). The comments state that feedback can only be used on even numbered LUTs but I'm struggling to see where this information comes from. 

I'm using ATtiny3227 and am about to try some things out to test this but wondered if @SpenceKonde would mind pointing me to the source of the limitation/freedom. I can see that the output of the sequencer is available as feedback but this is a no-sequencer setup (well at least the sequencer is not set up) and it looks like the feedback can go into either odd or even LUT, so I'm failing to follow along. 

> 28.3.1.4 Truth Table Inputs Selection
> Input Overview
> The inputs can be individually:
> • OFF
> • Driven by peripherals
> • Driven by internal events from the Event System
> • Driven by I/O pin inputs
> • Driven by other LUTs
> Internal Feedback Inputs (FEEDBACK)
> The output from a sequencer can be used as an input source for the two LUTs it is connected to.
> When selected (INSELy = FEEDBACK in LUTnCTRLx), the sequencer (SEQ) output is used as input for the
> corresponding LUTs.
![image](https://user-images.githubusercontent.com/26767525/179739227-98779712-245a-496f-9629-1816b3839d7f.png)

**Comment by SpenceKonde** [comment]
Yes, the feedback goes to both LUTs. But there is only ONE feedback path
per pair of LUTs, and it is the output of the sequencer if used, or if
sequencer is not used, it is the output of the even LU. Since that example
is about making a latch with only ONE CCL LUT, that means it can only be
done with the even lut, or by setting an event channel to use the odd lut
as generator, and then using that as an input. They omit mention of what th
feedback path does when the sequencer isn't in use in the datasheets, I
don't understand why

Personally, I found the sequencer features to be less than useful. Taking 2
luts hurts, because when you want them, you usually also want a lot more
LUTs too, and 90% of what they can be used to do can also be done with just
one LUT. (as the LatchNoSeq example shows).
They've got 6 sequencer operions, and 15 codepoints allocated to them, so
future devices could have more. I'd vote for a "X clock delay" where the
unused 4 high bits of the register set the delay, and an event extender
that worked the same way, to deal with those damned 1 system clock duration
pulse events
On the topic of "I wish the CCL/EVSYS did, why can't we use the PLL to
clock the LUTs?! They're fast enough - if you set one to use feedback as
it's only input, and invert it, it oscillates at around 100 MHz....  And
why can't we use the event system to supply the clock line for SPI other
than externally conncting pins (which a upcoming project will do - it's
part of my plan to output neopixel datastreams using SPI buffered mode, the
TCD  (dual ramp output compare override mode, such that it is both timing
the overall bit length, and outputting 1's on WOA and 0's on WOB. SPI in
slave mode with SCK, MISO and SS jumpered to other pins, timer generating
SCK, and a clock delayed longer than the setup time of MOSI, so when it
goes high, we lock in the value ot mosi and choose between WOA and WOB
being output on that LUT's pin. I recently calculated that it would give me
like 10x more clock cycles to fill the frame buffer.... Using Arduino
Random, I wasn't meeting frame rate targets or anything close on a 16 MHz
classic AVR running tinyNeoPixel_Static; ofc part of the problem is that
the random generator involves a 32-bit division, but still, I want higher
frame rate, 5x as many leds minimum, and and switching to an xorshift
algorithm would only help (albeit significantly) on the computationally
heavy patterns. And why is there no way to read the outputs of EVSYS from
software? they say each channel has a synchronizer on it already, all they
had to do is tie it to a register.

On Tue, Jul 19, 2022 at 7:28 AM SimonMerrett ***@***.***>
wrote:

> So this may not be in the correct category but I'm looking at an example
> from the logic library, specifically LatchNoSeq.ino
> <https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/libraries/Logic/examples/LatchNoSeq/LatchNoSeq.ino>.
> The comments state that feedback can only be used on even numbered LUTs but
> I'm struggling to see where this information comes from.
>
> I'm using ATtiny3227 and am about to try some things out to test this but
> wondered if @SpenceKonde <https://github.com/SpenceKonde> would mind
> pointing me to the source of the limitation/freedom. I can see that the
> output of the sequencer is available as feedback but this is a no-sequencer
> setup (well at least the sequencer is not set up) and it looks like the
> feedback can go into either odd or even LUT, so I'm failing to follow along.
>
> 28.3.1.4 Truth Table Inputs Selection
> Input Overview
> The inputs can be individually:
> • OFF
> • Driven by peripherals
> • Driven by internal events from the Event System
> • Driven by I/O pin inputs
> • Driven by other LUTs
> Internal Feedback Inputs (FEEDBACK)
> The output from a sequencer can be used as an input source for the two
> LUTs it is connected to.
> When selected (INSELy = FEEDBACK in LUTnCTRLx), the sequencer (SEQ) output
> is used as input for the
> corresponding LUTs.
> [image: image]
> <https://user-images.githubusercontent.com/26767525/179739227-98779712-245a-496f-9629-1816b3839d7f.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SpenceKonde/megaTinyCore/discussions/744>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABTXEWZGTLLSWKPODWO3UVLVU2GN3ANCNFSM537UAP2Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore <https://github.com/SpenceKonde/ATTinyCore>: Arduino support for
all pre-2016 tinyAVR with >2k flash!
megaTinyCore <https://github.com/SpenceKonde/megaTinyCore>: Arduino support
for all post-2016 tinyAVR parts!
DxCore <https://github.com/SpenceKonde/DxCore>: Arduino support for the AVR
Dx-series parts, the latest and greatest from Microchip!
Contact: ***@***.***

**Reply by SimonMerrett** [reply]
> it is the output of the sequencer if used...
> They omit mention of what the feedback path does when the sequencer isn't in use in the datasheets, I don't understand why...
Well done for finding this out without the help of Microchip

So, if my even LUT is spare but my ouput is tied to the odd LUT, I can still make this work. I just don't want to bother with the sequencer.

**Reply by SpenceKonde** [reply]
> Well done for finding this out without the help of Microchip

Microchip support has on not a single occasion been of any help to me. 

I discovered this when I was playing with using CCLs as clock prescalers for type B timers, now that newer parts can count on events. (see the tour de force in the oscillate example). It was almost immediately obvious what was happening.
